### PR TITLE
Bug 1565237 - Discover OAuth endpoints on console startup

### PR DESCRIPTION
**Pull request by spadgett**
/assign @deads2k 
cc @simo5 @jwforres 

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1565237

### CODE REVIEW

File: pkg/assets/apiserver/asset_apiserver.go [hunk 1]
```diff
@@ -105,13 +125,31 @@ func NewAssetServerConfig(config v1.WebConsoleConfiguration) (*AssetServerConfig
...
 	}
 
-	return cfg
+	restClient, err := kubernetes.NewForConfig(c.GenericConfig.ClientConfig)
+	if err != nil {
+		return completedConfig{}, err
+	}
+
+	// Discover the published OAuth endpoints from the well-known URL.
+	resultBytes, err := restClient.RESTClient().Get().AbsPath(oauthMetadataEndpoint).Do().Raw()
```
Comment (enj) [inline pkg/assets/apiserver/asset_apiserver.go:135]: Does this fail on all no 200 codes?  We had an issue with oc not erroring here so you would end up with an empty metadata struct.

File: pkg/assets/apiserver/asset_apiserver.go [hunk 2]
```diff
@@ -105,13 +125,31 @@ func NewAssetServerConfig(config v1.WebConsoleConfiguration) (*AssetServerConfig
...
+	if err != nil {
+		return completedConfig{}, err
+	}
+	metadata := &OAuthAuthorizationServerMetadata{}
+	if err := json.Unmarshal(resultBytes, metadata); err != nil {
+		return completedConfig{}, err
+	}
+	cfg.ExtraConfig.OAuthAuthorizationEndpoint = metadata.AuthorizationEndpoint
```
Comment (enj) [inline pkg/assets/apiserver/asset_apiserver.go]: Above comment code be addressed with checking both of these are not the empty string.

Comment (spadgett) [reply pkg/assets/apiserver/asset_apiserver.go]: @enj thanks, i added a check

File: pkg/assets/apiserver/asset_apiserver.go [hunk 3]
```diff
@@ -76,14 +87,24 @@ type AssetServer struct {
 
 type completedConfig struct {
 	GenericConfig genericapiserver.CompletedConfig
-	ExtraConfig   *ExtraConfig
+
+	// ClientConfig holds the kubernetes client configuration.
+	// This value is set by RecommendedOptions.CoreAPI.ApplyTo called by RecommendedOptions.ApplyTo.
+	// By default in-cluster client config is used.
+	ClientConfig *restclient.Config
```
Comment (deads2k) [inline pkg/assets/apiserver/asset_apiserver.go]: How does this value differ from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/config.go#L203 ?

File: pkg/assets/apiserver/asset_apiserver.go [hunk 4]
```diff
@@ -55,11 +58,19 @@ func init() {
...
+	// https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
+	oauthMetadataEndpoint = "/.well-known/oauth-authorization-server"
 )
 
 type ExtraConfig struct {
 	Options   v1.WebConsoleConfiguration
 	PublicURL url.URL
+
+	OAuthAuthorizationEndpoint string
+	OAuthTokenEndpoint         string
```
Comment (deads2k) [inline pkg/assets/apiserver/asset_apiserver.go:72]: I expected to see a polling loop in this pull that watches to see when the value changes and either live updates or suicides.  The value can change over time, same as anything else that can rotate.

**Comment by enj** [comment]
Spoke with David and Sam, we will do the remaining bits as a follow-up.

**Comment by spadgett** [comment]
Hm.

> During a build, a go file on disk could not be found: vendor/k8s.io/apimachinery/pkg/runtime/schema/generated.pb.go. It appears that files disappear from disk during a run - possibly due to some other component deleting them, or a temporary directory being cleaned up while a Go build is running.

/retest

File: pkg/assets/apiserver/asset_apiserver.go (hunk 3 above)
Comment (spadgett) [reply pkg/assets/apiserver/asset_apiserver.go]: Good catch. Fixed.

**Comment by spadgett** [comment]
Opened #36 for polling config.

**Pull request merged**